src/main/java/com/adobe/cq/testing/junit/rules/toggles/TogglesAwareTestRule.java

```diff
@@ -0,0 +1,121 @@
+import org.apache.sling.testing.clients.SlingClient;
+import org.apache.sling.testing.clients.util.poller.Polling;
+import org.junit.AssumptionViolatedException;
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+/**
+ * Junit rule for filtering tests based on the toggles enabled on the remote instance.<br>
+ * <br>
+ * The rule is used in combination with the annotations {@link RunIfToggleEnabled} and
+ * {@link SkipIfToggleEnabled}.
+ * If both annotations are applied to a test, both conditions must be met (AND operation).
+ *
+ * @see RunIfToggleEnabled
+ * @see SkipIfToggleEnabled
+ */
+public class TogglesAwareTestRule implements TestRule {
+    private static final Logger LOG = LoggerFactory.getLogger(TogglesAwareTestRule.class);
+
+    /**
+     * Supplier of the client needed for retrieving the enabled toggles.
+     */
+    private final Supplier<SlingClient> clientSupplier;
+
+    public TogglesAwareTestRule(Supplier<SlingClient> clientSupplier) {
+        this.clientSupplier = clientSupplier;
+    }
+
+    @Override
+    public Statement apply(Statement base, Description description) {
+        if (shouldRunTest(description)) {
+            return base;
+        } else {
+            return emptyStatement(description.getDisplayName());
+        }
+    }
+
+    protected boolean shouldRunTest(Description description) {
+        String runIfToggle = getRunIfToggle(description);
+        String skipIfToggle = getSkipIfToggle(description);
+
+        if (runIfToggle == null && skipIfToggle == null) {
+            return true; // no annotation was applied to the test
+        }
+
+        AtomicReference<List<String>> enabledToggles = new AtomicReference<>();
+        try {
+            new Polling(() -> {
+                enabledToggles.set(clientSupplier.get().adaptTo(TogglesClient.class).getEnabledToggles());
+                return true;
+            }).poll(SECONDS.toMillis(30), SECONDS.toMillis(1));
+        } catch (TimeoutException e) {
+            LOG.warn("Failed to retrieve toggles", e);
+            return true; // if something goes wrong we assume test should run
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            return true; // thread was interrupted, we assume test should run
+        }
+
+        if (runIfToggle != null && !enabledToggles.get().contains(runIfToggle)) {
+            return false; // RunIfToggleEnabled condition was not met, skipping
+        }
+
+        return skipIfToggle == null || !enabledToggles.get().contains(skipIfToggle);
+    }
+
+    private String getRunIfToggle(Description description) {
```

Comment:
minor: should we make this an `Optional<String>` ?

Reply:
👍 I've changed that in 03b7fa16517690ea71ef76b879c89ce02d728104. WDYT?